File: examples/tests/src/java/skywriting/examples/tests/java2/BinomialOptionsDriver.java
package skywriting.examples.tests.java2;

import java.io.DataInputStream;
import java.io.FileInputStream;

import com.asgow.ciel.executor.Ciel;
import com.asgow.ciel.references.Reference;
import com.asgow.ciel.tasks.FirstClassJavaTask;

public class BinomialOptionsDriver implements FirstClassJavaTask {

	@Override
	public Reference[] getDependencies() {
		return new Reference[0];
	}
	
	@Override
	public void invoke() throws Exception {
		
		int n = Integer.parseInt(Ciel.args[0]);
		int chunk_size = Integer.parseInt(Ciel.args[1]);
		boolean direct_fifos = Boolean.parseBoolean(Ciel.args[2]);
		
		double s = 100;
		double k = 100;
		double t = 1;
		double v = 0.3;
		double rf = 0.03;
		double cp = -1;
		
		FirstClassJavaTask lastNode = new BinomialOptions(null, s, k, t, v, rf, cp, n, chunk_size, true, direct_fifos);
	    Reference[] last_refs = Ciel.spawn(lastNode, new String[0], 1);
		for(int i = 0; i <= n; i+= chunk_size) {
			lastNode = new BinomialOptions(last_refs[0], s, k, t, v, rf, cp, n, chunk_size, false, direct_fifos);
			last_refs = Ciel.spawn(lastNode, new String[0], 1);
		}
		
		Ciel.blockOn(last_refs[0]);
		
		String filename = Ciel.RPC.getFilenameForReference(last_refs[0]);
		DataInputStream dis = new DataInputStream(new FileInputStream(filename));
		
		double ret = dis.readDouble();
		dis.close();
		
		Ciel.returnPlainString("The last node returned: " + ret + "\n");
		
	}

	@Override
	public void setup() {
		;
	}

}

File: examples/scala-tests/src/scala/BinomialOptionsDriver.scala

import java.io.DataInputStream;
import java.io.FileInputStream;

import com.asgow.ciel.executor.Ciel
import com.asgow.ciel.references.Reference
import com.asgow.ciel.references.CielFuture
import com.asgow.ciel.tasks.FirstClassJavaTask
import com.asgow.ciel.scala.SkylaFuture
import com.asgow.ciel.scala.SkylaThread

class BinomialOptionsDriver extends SkylaThread[String] {

      override def run = {
		
		val n = Ciel.args(0).toInt
		val chunk_size = Ciel.args(1).toInt
		
		val s: Double = 100
		val k: Double = 100
		val t: Double = 1
		val v: Double = 0.3
		val rf: Double = 0.03
		val cp: Double = -1
		
		var lastNode = new ScalaBinomialOptions(s, k, t, v, rf, cp, n, chunk_size, None)
	        var last_refs = Ciel.spawn(lastNode, new Array[String](0), 1)
		val range = scala.Range(0, n+1, chunk_size)
		for(i <- range) {
			lastNode = new ScalaBinomialOptions(s, k, t, v, rf, cp, n, chunk_size, Some(last_refs(0)))
			last_refs = Ciel.spawn(lastNode, new Array[String](0), 1)
		}
		
		val wait_ref = last_refs(0)
		val skyla_future = new SkylaFuture[Double](new CielFuture(wait_ref))

		val ret_double = skyla_future.get
		
		("The last node returned: " + ret_double + "\n")
		
	}


}

